linode/manager
# refactor: [M3-8623] - Introduce TanStack Router

![Screenshot 2024-10-03 at 11 45 08](https://github.com/user-attachments/assets/c456dcb8-390c-4806-b898-d691f66db5d3)

## Description 📝
This (large) PR introduces `TanStack Router` to Cloud Manager. 

https://tanstack.com/router/latest

It prepares a full migration from our old, outdated v5 `react-router-dom`  to the new Routing API.

The benefits of this migration are multiple:
- very low dependency
- 100% inferred TypeScript support
- Typesafe navigation
- Nested Routing and layout routes
- Built-in Route Loaders w/ SWR Caching
- Designed for client-side data caches (TanStack Query, SWR, etc.)
- Automatic route prefetching
- Asynchronous route elements and error boundaries
- File-based Route Generation
- Typesafe JSON-first Search Params state management APIs
- Path and Search Parameter Schema Validation
- Search Param Navigation APIs
- Custom Search Param parser/serializer support
- Search param middleware
- Route matching/loading middleware

> [!IMPORTANT]
> This initial PR does not (read **_should not_**) affect the current Cloud Manager routing experience. It only aims to achieve the following:
- consistent file splitting/lazy loading 
- centralized routes (parity with existing routes

I do not expect to have 100% route parity with this PR, I however did my very best get as close as possible. Subsequent PRs will rollout the actual new routing by updating legacy utils with the TanStack ones (params, location, history, etc)


## Changes  🔄
- Create lazy exports for routed components
- Create all existing routes in centralized location
- Configure root and routing architecture

## How to test 🧪

### Prerequisites
- pull code locally and run `yarn` to get new dependency

### Verification steps
- Confirm this PR does not introduce any functional change or regression
- Switch the routing in `<App />` to very basic functionality
  - **important** because we only defined the routes and have not replaced the utils, the experience may be very buggy for many routes - using MSW legacy handlers will help a bit with that matter


## As an Author I have considered 🤔

*Check all that apply*
